ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions.cpp

```diff
@@ -68,25 +69,68 @@ static Tensor reduce_impl(
         }
     }
 
+    auto transpose_and_compute_reduction = [&](const Tensor& input_tensor_arg,
+                                               const int dim1,
+                                               const int dim2,
+                                               const std::variant<int, ttnn::SmallVector<int>>& reduce_dim) -> Tensor {
+        Tensor output = ttnn::transpose(input_tensor_arg, dim1, dim2, memory_config);
+        output = reduce_impl<reduce_type>(
+            output, reduce_dim, /*keepdim=*/true, memory_config, compute_kernel_config, scalar, /*reshape=*/true);
+        output = ttnn::transpose(output, dim1, dim2, memory_config);
+        if (reshape) {
+            output = ttnn::reshape(output, ttnn::Shape{output_shape});
+        }
+        return output;
+    };
+
+    std::variant<int, ttnn::SmallVector<int>> reduce_dim;
     if (dim.size() == 1 && (rank == 3 || rank == 4)) {
         if (dim[0] == 1 && rank == 4) {
-            Tensor output = ttnn::transpose(input_tensor_arg, 1, -2, memory_config);
-            output = reduce_impl<reduce_type>(
-                output, 2, /*keepdim=*/true, memory_config, compute_kernel_config, scalar, /*reshape=*/true);
-            output = ttnn::transpose(output, 1, -2, memory_config);
-            if (reshape) {
-                output = ttnn::reshape(output, ttnn::Shape{output_shape});
-            }
-            return output;
+            reduce_dim = 2;
+            return transpose_and_compute_reduction(input_tensor_arg, 1, -2, reduce_dim);
         } else if (dim[0] == 0) {
-            Tensor output = ttnn::transpose(input_tensor_arg, 0, -2, memory_config);
-            output = reduce_impl<reduce_type>(
-                output, -2, /*keepdim=*/true, memory_config, compute_kernel_config, scalar, /*reshape=*/true);
-            output = ttnn::transpose(output, 0, -2, memory_config);
-            if (reshape) {
-                output = ttnn::reshape(output, ttnn::Shape{output_shape});
+            reduce_dim = -2;
+            return transpose_and_compute_reduction(input_tensor_arg, 0, -2, reduce_dim);
+        }
+    }
+
+    if (dim.size() == 2 && (rank == 3 || rank == 4)) {
```

Comment:
This may overlap with other existing code paths.
Also, special casing for each case is making the code too brittle and complex.

Please change the else clause that contains `TT_THROW("Unsupported dim");` to handle the logic, and figure out how to adapt the loops used for reducing on all dims.

Given those loops (that start with `for (int rank = ...`), it seems that calculating mean needs to be special cased and that var and std cannot be supported with the existing functionality. Probably because of non-linearity.

Ideally, you could combine all dim and multiple dim logic together into the current `TT_THROW("Unsupported dim");` clause, making code simpler as well.

Reply:
there is existing code which deals with dim_size =1 
`if (dim.size() == 1 && (rank == 3 || rank == 4))`

Hence i added for size=2 case as additional if condition.  Ill look into refactoring this code to address all combinations of input_sizes & dims